src/core/namespace.c

```diff
@@ -875,16 +875,55 @@ static bool namespace_info_mount_apivfs(const char *root_directory, const NameSp
         assert(ns_info);
 
         /*
-         * ProtectControlGroups= and ProtectKernelTunables= imply MountAPIVFS=,
-         * since to protect the API VFS mounts, they need to be around in the
-         * first place... and RootDirectory= or RootImage= need to be set.
+         * If one of ProtectControlGroups=, ProtectKernelTunables= or
+         * PrivateDevices= is set then MountAPIVFS= is implied, since to
+         * protect the API VFS mounts, they need to be around in the first
+         * place... and RootDirectory= or RootImage= need to be set.
          */
 
         /* root_directory should point to a mount point */
         return root_directory &&
                 (ns_info->mount_apivfs ||
                  ns_info->protect_control_groups ||
-                 ns_info->protect_kernel_tunables);
+                 ns_info->protect_kernel_tunables ||
+                 ns_info->private_dev);
+}
+
+static int append_mount_apivfs(MountEntry **p, const char *root_directory,
+                               const NameSpaceInfo *ns_info, bool ignore_protect) {
+        int r = 0;
+
+        assert(p);
+
+        if (!root_directory)
+                return 0;
+
+        /*
+         * If one of ProtectControlGroups=, ProtectKernelTunables= or PrivateDevices=
+         * are set, then mount the API file systems as they are in a protect
+         * mode.
+         */
+        if (ns_info->protect_control_groups ||
+            ns_info->protect_kernel_tunables || ns_info->private_dev) {
+                *((*p)++) = (MountEntry) {
+                        .path_const = "/sys",
+                        .mode = SYSFS,
+                        .ignore = ignore_protect,
+                };
+                *((*p)++) = (MountEntry) {
+                        .path_const = "/proc",
+                        .mode = PROCFS,
+                        .ignore = ignore_protect,
+                };
+                *((*p)++) = (MountEntry) {
+                        .path_const = "/dev",
+                        .mode = PRIVATE_DEV,
+                        .ignore = ignore_protect,
+                };
+        } else if (ns_info->mount_apivfs)
+                r = append_static_mounts(p, apivfs_table, ELEMENTSOF(apivfs_table), ignore_protect);
```

Comment:
The difference is in the `PRIVATE_DEV` and `BIND_DEV` , actually we can switch to use the same `append_static_mounts()` function and add the `PRIVATE_DEV` later based on those conditions, the `PRIVATE_DEV` will overwrite the `BIND_DEV`.

For why the change

Before:
* `MountAPIVFS=` was always implied if `RootDirectory=` and `ProtectControlGroups=`, or `ProtectKernelTunables=` 

 * `/dev` was a always a bind mount of host's /dev
 * `MountAPIVFS=yes` was *not* implied if `PrivateDevices=yes` and `RootDirectory=/...`


After:
* `MountAPIVFS=` is now always implied if `RootDirectory=` and `ProtectControlGroups=`, or  `ProtectKernelTunables=` *or*  `PrivateDevices=` 

* `MountAPIVFS=yes` is implied if `PrivateDevices=yes` and `RootDirectory=/...`.

* `/dev` now is a bind mount of host's /dev only if `ProtectControlGroups=` and `ProtectKernelTunables=` and `PrivateDevices=` are *not set*  otherwise it's a new private instance.

IMO if `ProtectKernelTunables=` was set under `RootImage=`  then don't make a bind mount of `/dev` where one can writes/reads to `/dev/mem` or other /dev nodes, read /proc/iomem or other kernel symbols, resolve addresses and go back to `/dev/*` which will defeat some `ProtectKernelTunables=` options. If the user sets `RootImage=` and `ProtectKernelTunables=` maybe he will be surprised if he finds out that `/dev` is a bind mount of host and present inside `RootImage=`.

Maybe the other solution is we should *explicitly* document it in `ProtectKernelTunables=` or `ProtectControlGroups=` doc, that you should also set `PrivateDevices=` ? in this case we can drop this patch...